### DISCUSSION TITLE
How can Junit 5 implement unit tests with multiple environments?

### DISCUSSION THREAD

**Discussion by sherlock-lin** [junit-team/junit-framework#3809] [Q&A]
In Junit 4, the project implements the execution of each unit test in multiple environments through @RunWith(Parameterized.class), @Parameterized.Parameter(0), and @Before. Now I want to upgrade this piece of logic to a Junit5 implementation, but I haven't found a more suitable way to transform it

**Comment by sherlock-lin** [comment]
In Junit 4, the following implementation allows each unit test to be executed in four different environments
```
    @Parameterized.Parameters
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][] {
            { true, true }, { true, false }, { false, true }, { false, false }
        });
    }

    @Parameterized.Parameter(0)
    public boolean useV2;

    @Parameterized.Parameter(1)
    public boolean writeJournal;

    @Before
    public void setUp() throws Exception {
        baseConf.setJournalWriteData(writeJournal);
        baseClientConf.setUseV2WireProtocol(useV2);

        super.setUp();
        rng = new Random(0); // Initialize the Random
        // Number Generator
        entries1 = new ArrayList<byte[]>(); // initialize the entries list
        entries2 = new ArrayList<byte[]>(); // initialize the entries list
    }
```

**Comment by sbrannen** [comment]
In JUnit Jupiter, you can write a `@ParameterizedTest` method (see [Parameterized Tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests) in the User Guide).

Please note, however, that `@ParameterizedTest` can only be used at the method level.

Support for class-level parameterization has not yet been implemented in JUnit Jupiter (see #878).

**Comment by mpkorstanje** [comment]
A rather hacky work around would be create an JUnit Jupiter extension that reads configuration parameters and provides them to tests. Then you can use the JUnit Platform Suite Engine to create a suite for each environment, each suite configured with a a specific set of configuration parameters for that environment.
